src/V3LinkInc.cpp

```diff
@@ -0,0 +1,239 @@
+            // Add the whole thing before insertAt
+            if (debug() >= 9) newp->dumpTree(cout, "-newfunc:");
+            m_insStmtp->addHereThisAsNext(newp);
+        } else if (m_insMode == IM_AFTER) {
+            m_insStmtp->addNextHere(newp);
+        } else if (m_insMode == IM_WHILE_PRECOND) {
+            AstWhile* whilep = VN_CAST(m_insStmtp, While);
+            UASSERT_OBJ(whilep, nodep, "Insert should be under WHILE");
+            whilep->addPrecondsp(newp);
+        } else {
+            nodep->v3fatalSrc("Unknown InsertMode");
+        }
+        m_insMode = IM_AFTER;
+        m_insStmtp = newp;
+    }
+
+    // VISITORS
+    virtual void visit(AstWhile* nodep) VL_OVERRIDE {
+        // Special, as statements need to be put in different places
+        // Preconditions insert first just before themselves (the normal
+        // rule for other statement types)
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->precondsp());
+        // Conditions insert first at end of precondsp.
+        m_insMode = IM_WHILE_PRECOND;
+        m_insStmtp = nodep;
+        iterateAndNextNull(nodep->condp());
+        // Body insert just before themselves
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->bodysp());
+        iterateAndNextNull(nodep->incsp());
+        // Done the loop
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+    virtual void visit(AstNodeFor* nodep) VL_OVERRIDE {
+        nodep->v3fatalSrc(
+            "For statements should have been converted to while statements in V3Begin.cpp");
+    }
+    virtual void visit(AstNodeStmt* nodep) VL_OVERRIDE {
+        if (!nodep->isStatement()) {
+            iterateChildren(nodep);
+            return;
+        }
+        m_insMode = IM_BEFORE;
+        m_insStmtp = nodep;
+        iterateChildren(nodep);
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+    void unsupported_visit(AstNode* nodep) {
+        m_unsupportedHere = true;
+        iterateChildren(nodep);
+        m_unsupportedHere = false;
+    }
+    virtual void visit(AstLogAnd* nodep) VL_OVERRIDE { unsupported_visit(nodep); }
+    void prepost_visit(AstNode* nodep) {
+        // Check if we are underneath a statement
+        if (m_insStmtp == NULL) {
+            prepost_non_stmt_visit(nodep);
+        } else {
+            prepost_stmt_visit(nodep);
+        }
+    }
+    void prepost_non_stmt_visit(AstNode* nodep) {
+        iterateChildren(nodep);
+
+        AstNRelinker linker;
+        AstNode* innerp = nodep->op1p();
+        AstNode* inner_rhsp = innerp->cloneTree(true);
+        AstConst* constp = VN_CAST(nodep->op2p(), Const);
+        AstConst* newconstp = constp->cloneTree(true);
+
+        // Make sure it can be lvalue if it's a VarRef
+        if (VN_IS(innerp, VarRef)) {
+            AstNodeVarRef* vr = VN_CAST(innerp, VarRef);
```

Comment:
```suggestion
        if (AstNodeVarRef* vr = VN_CAST(innerp, VarRef)) {
```

Reply:
Bigger issue is this won't properly lvalue the lvalue below.  You need to improve V3LinkLValue to properly handle these new node types, then you can remove this code section.

Reply:
But I am only setting the new `VarRef` to be lvalue.
I.e from `vr++` to `new_vw = vr + 1`, etc. I am not setting the original `vr` to be lvalue.

I am not sure if that's possible as this copy is created here (a few lines above) and V3LinkLValue happens earlier.

Reply:
Oh, right.  I still think you might have a problem where it doesn't get set if the varref is way below  (e.g. your if() is false).  E.g. an unpacked array foo[1][2][3]++.  You may need to iterate down in innerp to set it.

Reply:
Hmm, thanks for bringing that up, I think `foo[1][2][3]++` may not currently work at all, not only because the lvalue is not set. I think it needs to be handled separately in various places (where we don't have VarRef directly).

I need to rethink parts of the current implementation.